### PR TITLE
smarter logical combination that keeps truthy/falsy/nullish property

### PR DESCRIPTION
**Pull request by sokra**
### Description

Allows to handle `if (unknown && falsy)` `if (unknown || truthy)` `if (unknown ?? nullish)` in static analysis.

Hint: review each commit on it's own, since the first commit is only changing to a named struct.

### CODE REVIEW

**Comment by ForsakenHarmony** [comment]
> Allows to handle `if (unknown && falsy)` `if (unknown || truthy)` `if (unknown ?? nullish)` in static analysis.

Which cases were handled before this PR?
Just `if (falsy)` `if (truthy)` `if (nullish)`?

Does this also handle `if (unknown && falsy && unknown)` etc.?

File: crates/turbopack-ecmascript/src/analyzer/mod.rs [hunk 1]
```diff
@@ -350,6 +350,25 @@ enum JsValueMetaKind {
     Placeholder,
 }
 
+#[derive(Debug, Clone, Copy, Hash, PartialEq, Eq)]
+pub enum AdditionalProperty {
```
Comment (ForsakenHarmony) [inline crates/turbopack-ecmascript/src/analyzer/mod.rs:354]: Not a fan of this name, it's a bit too generic

**Comment by sokra** [comment]
> Which cases were handled before this PR?
> Just `if (falsy)` `if (truthy)` `if (nullish)`?

The `unknown` was the problem. If was able to handle `if (truthy || unknown)` but not `if (unknown || truthy)`.
The problem is that `truthy || unknown => truthy`, but `unknown || truthy` could be `unknown` (if it's truthy at runtime) or `truthy`, so the evaluator evaluates it to `Alternatives(unknown, truthy)`. The PR adds an additional info to `Alternatives` that the combination has a truthy value.

File: crates/turbopack-ecmascript/src/analyzer/mod.rs (hunk 1 above)
Comment (sokra) [reply crates/turbopack-ecmascript/src/analyzer/mod.rs]: yeah, I changed it

**Comment by sokra** [comment]
* **#8890** <a href="https://app.graphite.dev/github/pr/vercel/turbo/8890?utm_source=stack-comment-icon" target="_blank">Graphite</a>
* **#8873** <a href="https://app.graphite.dev/github/pr/vercel/turbo/8873?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @sokra and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: crates/turbopack-ecmascript/src/analyzer/builtin.rs [hunk 2]
```diff
@@ -501,43 +503,100 @@ pub fn replace_builtin(value: &mut JsValue) -> bool {
...
                     }
                     None => {
                         // We don't know if this part is skipped or the final value, so we keep it.
+                        part_properties.push(property);
                         parts.push(part);
                         continue;
                     }
                 }
             }
+            part_properties.truncate(parts.len());
```
Comment (arlyon) [inline crates/turbopack-ecmascript/src/analyzer/builtin.rs:540]: under what condition does this happen? it seems from the code that part properties is always less than or equal to parts

Comment (sokra) [reply crates/turbopack-ecmascript/src/analyzer/builtin.rs:540]: never, removed it...

**Review by ForsakenHarmony — APPROVED** [review]

**Pull request merged**